core/src/parking_lot.rs

```diff
@@ -1035,3 +1042,67 @@ unsafe fn unpark_filter_internal(key: usize,
 
     result
 }
+
+/// Register an resource identified by key in the deadlock detector
+#[cfg(feature = "deadlock_detection")]
+pub fn register_resource(key: usize) {
+    unsafe {
+        let mut thread_data = None;
+        let thread_data = get_thread_data(&mut thread_data);
+        (*thread_data.resources.get()).push(key);
+    }
+}
+
+/// Deregister an resource identified by key in the deadlock detector
+#[cfg(feature = "deadlock_detection")]
+pub fn deregister_resource(key: usize) {
+    unsafe {
+        let mut thread_data = None;
+        let thread_data = get_thread_data(&mut thread_data);
+        let resources = &mut (*thread_data.resources.get());
+        match resources.iter().rposition(|x| *x == key) {
+            Some(p) => resources.swap_remove(p),
+            None => {
+                check_deadlock();
+                panic!("key {} not found in thread resources", key);
+            },
+        };
+    }
+}
+
+/// Returns true if a deadlock is detected
+#[cfg(feature = "deadlock_detection")]
+pub fn check_deadlock() -> bool {
+    // basic algorithm that builds a directed graph of the threads and the resources
+    // then checks for the presence of cycles (deadlocks)
```

Comment:
While checking for deadlocks, you should lock the entire hash table, not just one bucket at a time. See `grow_hashtable` for how to do this properly. You might get false positives if you do not do this.

Reply:
I forgot to comment about this. I was thinking that the actual implementation could do a first-run like this and then do a second pass to confirm (locking all queues).